### PR TITLE
cross validation with purge and embargo + walkforward

### PR DESCRIPTION
**Pull request by mjvakili**

### CODE REVIEW

File: quantfinlib/train_test_split/kfoldpe.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            train_index = _purge(
+                train_index=train_index, test_index=test_index, indices=indices, groups=groups, n_purge=self.n_purge
+            )
+            train_index = _embargo(
+                train_index=train_index, test_index=test_index, indices=indices, groups=groups, n_embargo=self.n_embargo
+            )
+            yield train_index, test_index
+
+
+def modify_fold_bnds(bnd_per_fold: np.ndarray, groups: np.ndarray) -> np.ndarray:
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:138]: maybe `bounds` or `limits`? `bnds` is not unambiguous.

File: quantfinlib/train_test_split/_base.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+from abc import ABC, abstractmethod
+from numbers import Integral
+from typing import Any, Optional, Union
+
+import numpy as np
+import pandas as pd
+
+
+def _validate_purge_embargo_inputs(
+    train_index: np.ndarray,
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/_base.py:12]: not sure if it is an overoptimization, but np.ndarray is repeated so many times here. If there is also a condition on dimentionality, then it maybe worth creating a custom type using https://github.com/ramonhagenaars/nptyping 

from nptyping import NDArray, Float
np1d = NDArray[Shape["*"], Float]
np2d = NDArray[Shape["*, *"], Float]

and then use it across the library.

File: quantfinlib/train_test_split/walk_forward.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+class WalkForward(BaseCV):
+    """Walk forward train-test split."""
+
+    def __init__(
+        self,
+        n_embargo: Union[Integral, pd.Timedelta],
+        n_purge: Union[Integral, pd.Timedelta] = 0,
+        train_window: Union[Integral, pd.Timedelta] = 0,
+        test_window: Union[Integral, pd.Timedelta] = 0,
+        wf_split_type: str = "expanding",
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/walk_forward.py:21]: split_type

File: quantfinlib/train_test_split/walk_forward.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        self.train_window = train_window
+        self.test_window = test_window
+        self.wf_split_type = wf_split_type
+        if wf_split_type not in ["expanding", "rolling"]:
+            raise ValueError("Invalid value for 'wf_split_type'. Must be either 'expanding' or 'rolling'.")
+        self.freq = freq
+        self.n_splits = 0
+        self._finalized_splits = False
+
+    def __repr__(self):
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/walk_forward.py:34]: I remember that __repr__ is for developers, __str__ is for users, but in case there is no __str__ any call of str(obj) will fall back to __repr__, so it is fine.

File: quantfinlib/train_test_split/kfoldpe.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    Parameters
+    ----------
+    indices : np.ndarray
+        All indices.
+    test_index : np.ndarray
+        Test indices.
+    groups : Union[np.ndarray, pd.TimedeltaIndex]
+        Groups.
+    look_forward : bool
+        Whether to test set is always ahead of the training set.
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:186]: ...THE test set...

File: quantfinlib/train_test_split/kfoldpe.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        last_group_curr_fold = groups[bnd_per_fold[fold, 1]]
+        first_group_next_fold = groups[bnd_per_fold[fold + 1, 0]]
+        if last_group_curr_fold == first_group_next_fold:
+            next_group = groups[groups > last_group_curr_fold][0]
+            bnd_per_fold[fold, 1] = np.where(groups == last_group_curr_fold)[0][-1]
+            bnd_per_fold[fold + 1, 0] = np.where(groups == next_group)[0][0]
+    return bnd_per_fold
+
+
+def get_train_index_per_split(
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:169]: get_train_index_per_split and modify_fold_bnds are both utility func for internal use, right? should start with _ then?

File: quantfinlib/train_test_split/kfoldpe.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    Attributes
+    ----------
+    n_folds : int
+        Number of splits.
+    n_purge : Union[int, pd.Timedelta]
+        Number of groups to purge.
+    n_embargo : Union[int, pd.Timedelta]
+        Number of groups to embargo.
+    look_forward : bool
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:40]: explanation is not clear, but get_train_index_per_split has a better one:
  Whether the test set is always ahead of the training set.

File: quantfinlib/train_test_split/kfoldpe.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        Groups.
+    look_forward : bool
+        Whether to test set is always ahead of the training set.
+
+    Returns
+    -------
+    np.ndarray
+        Train indices for a given split.
+    """
+    train_index = np.setdiff1d(indices, test_index)
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:193]: now train_index is computed twice in case look_forward is True

File: quantfinlib/train_test_split/kfoldpe.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+
+class TimeAwareKFold(BaseCV):
+    """Purge-embargo K-Fold cross-validator.
+
+    Provides train/test indices to split data in train/test sets. The training set is purged and embargoed.
+
+    Attributes
+    ----------
+    n_folds : int
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/kfoldpe.py:34]: n_folds parameter, but then you use `splits` later on in varialbe names and docstrings. also WalkForward uses `splits` notation.

File: quantfinlib/train_test_split/walk_forward.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            f"train_window={self.train_window}, test_window={self.test_window}, "
+            f"wf_split_type={self.wf_split_type})"
+        )
+
+    def get_n_splits(self) -> int:
+        """Return the number of splits."""
+        if self._finalized_splits:
+            return self.n_splits
+        else:
+            raise ValueError("Splits have not been finalized. Call the split method first.")
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/walk_forward.py:47]: maybe a warning instead of an error here?

File: quantfinlib/train_test_split/walk_forward.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+from typing import Generator, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+
+from quantfinlib.train_test_split._base import BaseCV, _embargo
+
+
+class WalkForward(BaseCV):
+    """Walk forward train-test split."""
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/walk_forward.py:13]: missing docs

File: quantfinlib/train_test_split/_base.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    _validate_purge_embargo_inputs(train_index, test_index, indices, groups, n_embargo)
+    min_test_group = groups[test_index].min()
+    embargoed_index = indices[(groups >= min_test_group - n_embargo) & (groups < min_test_group)]
+    train_index = np.setdiff1d(train_index, embargoed_index)
+    if len(train_index) == 0:
+        raise ValueError("All training indices were embargoed. No training data available.")
+    return train_index
+
+
+class BaseCV(ABC):
```
Comment (fluentnumbers) [inline quantfinlib/train_test_split/_base.py:134]: would be super cool to somehow visualize the split result. to see what was embargoed, purged, etc. because you yield a generator and not full result at once, it may not be feasible until fully yielded, I am not sure.

Comment (fluentnumbers) [inline quantfinlib/train_test_split/_base.py:134]: If there are more CV methods coming, it can be upgraded to the Strategy design pattern: https://refactoring.guru/design-patterns/strategy/python/example 

it only needs a CV_context(split_method:BaseCV) class with methods like: set_split_method() and split().
then it can be used as in:

cv = CV_context(kfold)
x,y = cv.split()
cv.set_method(wf)
x,y = cv.split()

File: tests/train_test_split/test_base.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    group_indices = np.random.randint(0, 60, 100)
+    # make sure that all groups between 0 and 32 are present in the data
+    group_indices = np.unique(group_indices, return_inverse=True)[1].reshape(group_indices.shape)
+    group_indices.sort()
+    return groups[group_indices]
+
+
+@pytest.fixture(
+    params=[
+        ("integer", np.arange(100), integer_even_groups()),
```
Comment (fluentnumbers) [inline tests/train_test_split/test_base.py:67]: I wanted to run tests with profiling on larger arrays to see where are the processing bottlenecks, but it is not so easy as there are many places where 100 needs to be changed to million or smth. can we define some variables on top to easily control testing parameters? I understand, this is probably not the point of unit-tests...

File: quantfinlib/train_test_split/kfoldpe.py (hunk 1 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:138]: Sounds good. fixed this

File: quantfinlib/train_test_split/walk_forward.py (hunk 3 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/walk_forward.py:21]: fixed

File: quantfinlib/train_test_split/kfoldpe.py (hunk 5 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:186]: fixed

File: quantfinlib/train_test_split/kfoldpe.py (hunk 6 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:169]: good suggestion. made them private functions.

File: quantfinlib/train_test_split/kfoldpe.py (hunk 7 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:40]: Made it clear.

File: quantfinlib/train_test_split/kfoldpe.py (hunk 8 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:193]: fixe it.

File: quantfinlib/train_test_split/kfoldpe.py (hunk 9 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/kfoldpe.py:34]: I made it clear in the docstring. n_splits is n_folds if look_forward is False, otherwise it is n_folds - 1 because we remove discard the train-test split for which the test set is behind the train set time-wise.

File: quantfinlib/train_test_split/walk_forward.py (hunk 11 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/walk_forward.py:13]: Added the docstring.

File: quantfinlib/train_test_split/_base.py (hunk 12 above)
Comment (mjvakili) [reply quantfinlib/train_test_split/_base.py:134]: There will be only one more CV method: combinatorial purged cross-validation. So, I don't think we need to change the design.

[the remark this answers is not in the document]
Comment (mjvakili) [reply quantfinlib/train_test_split/_base.py:134]: It is consistent with sklearn's BaseCrossValidator in that it implements the abstract method get_n_splits and the method split(X, y, groups). In that sense, you can supply these splitters to a GridSearchCV or optuna gridsearch.

Comment (mjvakili) [reply quantfinlib/train_test_split/_base.py:136]: Added an example notebook.

File: tests/train_test_split/test_base.py (hunk 13 above)
Comment (mjvakili) [reply tests/train_test_split/test_base.py:67]: I improved this in the unittests.
